File: examples/sdk/goose-sdk.ts
/**
 * goose TypeScript/JavaScript SDK
 * 
 * A TypeScript SDK for integrating goose AI agent into your applications.
 * 
 * Example usage:
 * ```typescript
 * import { GooseClient } from './goose-sdk';
 * 
 * const client = new GooseClient('http://localhost:3001', 'your-secret-key');
 * 
 * // Start agent
 * await client.startAgent({ provider: 'openai', model: 'gpt-4' });
 * 
 * // Send message and get response
 * for await (const event of client.sendMessage('my-session', 'Hello!')) {
 *   if (event.message) {
 *     console.log(event.message);
 *   }
 * }
 * ```
 */

export interface MessageContent {
  type: string;
  text: string;
  annotations?: Record<string, any>;
}

export interface Message {
  role: string;
  created: number;
  content: MessageContent[];
}

export interface Event {
  type: string;
  message?: Message;
  raw?: Record<string, any>;
}

export interface Extension {
  name: string;
  type: string;
  cmd?: string;
  args?: string[];
  envs?: Record<string, string>;
  timeout?: number;
  bundled?: boolean;
}

export interface Session {
  id: string;
  name?: string;
  created?: number;
  updated?: number;
  working_dir?: string;
}

export interface StartAgentOptions {
  provider?: string;
  model?: string;
  working_dir?: string;
}

export interface SendMessageOptions {
  working_dir?: string;
}

/**
 * Client for interacting with the goose HTTP API.
 */
export class GooseClient {
  private baseUrl: string;
  private secretKey: string;
  private timeout: number;

  /**
   * Create a new goose client.
   * 
   * @param baseUrl - Base URL of the goose server (default: http://localhost:3001)
   * @param secretKey - Authentication secret key (default: "test")
   * @param timeout - Request timeout in milliseconds (default: 60000)
   */
  constructor(
    baseUrl: string = 'http://localhost:3001',
    secretKey: string = 'test',
    timeout: number = 60000
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.secretKey = secretKey;
    this.timeout = timeout;
  }

  /**
   * Get default headers with authentication.
   */
  private getHeaders(additionalHeaders?: Record<string, string>): Record<string, string> {
    return {
      'X-Secret-Key': this.secretKey,
      'Content-Type': 'application/json',
      ...additionalHeaders,
    };
  }

  /**
   * Make an HTTP request to the goose API.
   */
  private async request<T = any>(
    method: string,
    path: string,
    options?: RequestInit
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const headers = this.getHeaders(options?.headers as Record<string, string>);

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(url, {
        method,
        headers,
        signal: controller.signal,
        ...options,
      });

      if (!response.ok) {
        throw new Error(
          `Request failed: ${response.status} ${response.statusText}`
        );
      }

      const contentType = response.headers.get('content-type');
      if (contentType?.includes('application/json')) {
        return await response.json();
      }
      return (await response.text()) as any;
    } finally {
      clearTimeout(timeoutId);
    }
  }

  // Agent Management

  /**
   * Start a new agent session.
   */
  async startAgent(options: StartAgentOptions = {}): Promise<any> {
    const payload = {
      provider: options.provider || 'openai',
      model: options.model || 'gpt-4',
      ...(options.working_dir && { working_dir: options.working_dir }),
    };

    return this.request('POST', '/agent/start', {
      body: JSON.stringify(payload),
    });
  }

  /**
   * Resume an existing agent session.
   */
  async resumeAgent(sessionId: string): Promise<any> {
    return this.request('POST', '/agent/resume', {
      body: JSON.stringify({ session_id: sessionId }),
    });
  }

  /**
   * Update the agent's LLM provider and model.
   */
  async updateAgentProvider(provider: string, model: string): Promise<any> {
    return this.request('POST', '/agent/update_provider', {
      body: JSON.stringify({ provider, model }),
    });
  }

  // Messaging

  /**
   * Send a message to the agent and stream the response.
   */
  async *sendMessage(
    sessionId: string,
    message: string,
    options: SendMessageOptions = {}
  ): AsyncGenerator<Event> {
    const payload = {
      messages: [
        {
          role: 'user',
          created: Math.floor(Date.now() / 1000),
          content: [{ type: 'text', text: message }],
        },
      ],
      session_id: sessionId,
      session_working_dir: options.working_dir || '.',
    };

    const response = await fetch(`${this.baseUrl}/reply`, {
      method: 'POST',
      headers: this.getHeaders({ Accept: 'text/event-stream' }),
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      throw new Error(`Request failed: ${response.status} ${response.statusText}`);
    }

    const reader = response.body!.getReader();
    const decoder = new TextDecoder();
    let buffer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          try {
            const data = JSON.parse(line.slice(6));
            const event: Event = {
              type: data.type || 'Unknown',
              raw: data,
            };

            if (data.type === 'Finish') {
              yield event;
              return;
            }

            if (data.message) {
              const msgData = data.message;
              const content: MessageContent[] = (msgData.content || []).map(
                (c: any) => ({
                  type: c.type,
                  text: c.text || '',
                  annotations: c.annotations,
                })
              );

              event.message = {
                role: msgData.role,
                created: msgData.created,
                content,
              };
            }

            yield event;
          } catch (e) {
            // Skip invalid JSON
            continue;
          }
        }
      }
    }
  }

  // Extension Management

  /**
   * Get list of available tools.
   */
  async getTools(): Promise<any[]> {
    return this.request('GET', '/agent/tools');
  }

  /**
   * Add an extension to the agent.
   */
  async addExtension(extension: Extension): Promise<string> {
    const payload: Record<string, any> = {
      name: extension.name,
      type: extension.type,
    };

    if (extension.cmd) payload.cmd = extension.cmd;
    if (extension.args) payload.args = extension.args;
    if (extension.envs) payload.envs = extension.envs;
    if (extension.timeout) payload.timeout = extension.timeout;
    if (extension.bundled !== undefined) payload.bundled = extension.bundled;

    return this.request('POST', '/agent/add_extension', {
      body: JSON.stringify(payload),
    });
  }

  /**
   * Remove an extension from the agent.
   */
  async removeExtension(name: string): Promise<string> {
    return this.request('POST', '/agent/remove_extension', {
      body: JSON.stringify({ name }),
    });
  }

  /**
   * List all available extensions.
   */
  async listExtensions(): Promise<any[]> {
    return this.request('GET', '/config/extensions');
  }

  // Session Management

  /**
   * List all sessions.
   */
  async listSessions(): Promise<Session[]> {
    const data = await this.request<{ sessions: any[] }>('GET', '/sessions');
    return data.sessions.map((s: any) => ({
      id: s.id,
      name: s.name,
      created: s.created,
      updated: s.updated,
      working_dir: s.working_dir,
    }));
  }

  /**
   * Get detailed session information.
   */
  async getSession(sessionId: string): Promise<any> {
    return this.request('GET', `/sessions/${sessionId}`);
  }

  /**
   * Update a session's name.
   */
  async updateSessionName(sessionId: string, name: string): Promise<void> {
    await this.request('POST', `/sessions/${sessionId}/name`, {
      body: JSON.stringify({ name }),
    });
  }

  /**
   * Delete a session.
   */
  async deleteSession(sessionId: string): Promise<void> {
    await this.request('DELETE', `/sessions/${sessionId}`);
  }

  /**
   * Export a session's data.
   */
  async exportSession(sessionId: string): Promise<any> {
    return this.request('GET', `/sessions/${sessionId}/export`);
  }

  /**
   * Get insights about all sessions (token usage, costs, etc.).
   */
  async getSessionInsights(): Promise<any> {
    return this.request('GET', '/sessions/insights');
  }

  // Configuration

  /**
   * Get list of available LLM providers.
   */
  async getProviders(): Promise<any[]> {
    return this.request('GET', '/config/providers');
  }

  /**
   * Get available models for a provider.
   */
  async getProviderModels(provider: string): Promise<string[]> {
    return this.request('GET', `/config/providers/${provider}/models`);
  }

  /**
   * Check server status.
   */
  async checkStatus(): Promise<any> {
    return this.request('GET', '/status');
  }
}

// Example usage
async function example() {
  const client = new GooseClient();

  // Check status
  const status = await client.checkStatus();
  console.log('Server status:', status);

  // Start agent
  console.log('\nStarting agent...');
  await client.startAgent({ provider: 'openai', model: 'gpt-4' });

  // List available extensions
  const extensions = await client.listExtensions();
  console.log(`\nAvailable extensions: ${extensions.length}`);

  // Send message
  console.log('\nSending message...');
  const sessionId = 'example-session';

  for await (const event of client.sendMessage(
    sessionId,
    'Write a simple hello world JavaScript function'
  )) {
    if (event.type === 'text' && event.message) {
      for (const content of event.message.content) {
        if (content.type === 'text') {
          console.log(`\nAgent: ${content.text}`);
        }
      }
    } else if (event.type === 'Finish') {
      console.log('\nConversation finished');
    }
  }
}

// Run example if this file is executed directly
if (require.main === module) {
  example().catch(console.error);
}

File: examples/sdk/goose_sdk.py
"""
goose Python SDK

A simple Python SDK for integrating goose AI agent into your applications.

Example usage:
    from goose_sdk import GooseClient
    
    async with GooseClient("http://localhost:3001", "your-secret-key") as client:
        # Start agent
        await client.start_agent(provider="openai", model="gpt-4")
        
        # Send message and get response
        async for event in client.send_message("my-session", "Hello!"):
            if event.message:
                print(event.message)
"""

import asyncio
import json
from dataclasses import dataclass
from datetime import datetime
from typing import AsyncIterator, Dict, Any, List, Optional, Union, Tuple
import httpx


@dataclass
class MessageContent:
    """Content of a message."""
    type: str
    text: str
    annotations: Optional[Dict[str, Any]] = None


@dataclass
class Message:
    """A message in the conversation."""
    role: str
    created: int
    content: List[MessageContent]


@dataclass
class Event:
    """An event from the agent response stream."""
    type: str
    message: Optional[Message] = None
    raw: Optional[Dict[str, Any]] = None


@dataclass
class Extension:
    """Extension configuration."""
    name: str
    type: str
    cmd: Optional[str] = None
    args: Optional[List[str]] = None
    envs: Optional[Dict[str, str]] = None
    timeout: Optional[int] = None
    bundled: Optional[bool] = None


@dataclass
class Session:
    """Session information."""
    id: str
    name: Optional[str] = None
    created: Optional[int] = None
    updated: Optional[int] = None
    working_dir: Optional[str] = None


class GooseClient:
    """
    Client for interacting with the goose HTTP API.
    
    Args:
        base_url: Base URL of the goose server (default: http://localhost:3001)
        secret_key: Authentication secret key (default: "test")
        timeout: Request timeout in seconds (default: 60.0)
    """
    
    # Default timeout for streaming operations
    STREAMING_TIMEOUT = 300.0
    
    def __init__(
        self,
        base_url: str = "http://localhost:3001",
        secret_key: str = "test",
        timeout: float = 60.0
    ):
        self.base_url = base_url.rstrip('/')
        self.secret_key = secret_key
        self.timeout = timeout
        self._client: Optional[httpx.AsyncClient] = None
    
    async def __aenter__(self):
        self._client = httpx.AsyncClient(timeout=self.timeout)
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self._client:
            await self._client.aclose()
    
    def _headers(self) -> Dict[str, str]:
        """Get default headers with authentication."""
        return {
            "X-Secret-Key": self.secret_key,
            "Content-Type": "application/json"
        }
    
    async def _request(
        self,
        method: str,
        path: str,
        **kwargs
    ) -> httpx.Response:
        """Make an HTTP request to the goose API."""
        if not self._client:
            raise RuntimeError("Client not initialized. Use 'async with' context manager.")
        
        url = f"{self.base_url}{path}"
        headers = self._headers()
        if 'headers' in kwargs:
            headers.update(kwargs.pop('headers'))
        
        response = await self._client.request(
            method,
            url,
            headers=headers,
            **kwargs
        )
        response.raise_for_status()
        return response
    
    # Agent Management
    
    async def start_agent(
        self,
        provider: str = "openai",
        model: str = "gpt-4",
        working_dir: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Start a new agent session.
        
        Args:
            provider: LLM provider name (e.g., "openai", "anthropic")
            model: Model name (e.g., "gpt-4", "claude-3-opus")
            working_dir: Working directory for the agent
            
        Returns:
            Agent initialization response
        """
        payload = {
            "provider": provider,
            "model": model
        }
        if working_dir:
            payload["working_dir"] = working_dir
        
        response = await self._request("POST", "/agent/start", json=payload)
        return response.json()
    
    async def resume_agent(self, session_id: str) -> Dict[str, Any]:
        """
        Resume an existing agent session.
        
        Args:
            session_id: ID of the session to resume
            
        Returns:
            Agent resume response
        """
        response = await self._request(
            "POST",
            "/agent/resume",
            json={"session_id": session_id}
        )
        return response.json()
    
    async def update_agent_provider(
        self,
        provider: str,
        model: str
    ) -> Dict[str, Any]:
        """
        Update the agent's LLM provider and model.
        
        Args:
            provider: New provider name
            model: New model name
            
        Returns:
            Update response
        """
        response = await self._request(
            "POST",
            "/agent/update_provider",
            json={"provider": provider, "model": model}
        )
        return response.json()
    
    # Messaging
    
    async def send_message(
        self,
        session_id: str,
        message: str,
        working_dir: str = "."
    ) -> AsyncIterator[Event]:
        """
        Send a message to the agent and stream the response.
        
        Args:
            session_id: Session ID
            message: Message text
            working_dir: Working directory for the session
            
        Yields:
            Event objects from the agent response stream
        """
        if not self._client:
            raise RuntimeError("Client not initialized. Use 'async with' context manager.")
        
        payload = {
            "messages": [{
                "role": "user",
                "created": int(datetime.now().timestamp()),
                "content": [{"type": "text", "text": message}]
            }],
            "session_id": session_id,
            "session_working_dir": working_dir
        }
        
        headers = self._headers()
        headers["Accept"] = "text/event-stream"
        
        async with self._client.stream(
            "POST",
            f"{self.base_url}/reply",
            json=payload,
            headers=headers,
            timeout=self.STREAMING_TIMEOUT
        ) as stream:
            async for line in stream.aiter_lines():
                if not line:
                    continue
                
                if line.startswith("data: "):
                    try:
                        data = json.loads(line[6:])
                        event = Event(type=data.get("type", "Unknown"), raw=data)
                        
                        if data.get("type") == "Finish":
                            yield event
                            break
                        
                        if "message" in data:
                            msg_data = data["message"]
                            content = []
                            for c in msg_data.get("content", []):
                                content.append(MessageContent(
                                    type=c.get("type"),
                                    text=c.get("text", ""),
                                    annotations=c.get("annotations")
                                ))
                            event.message = Message(
                                role=msg_data.get("role"),
                                created=msg_data.get("created"),
                                content=content
                            )
                        
                        yield event
                    except json.JSONDecodeError:
                        continue
    
    # Extension Management
    
    async def get_tools(self) -> List[Dict[str, Any]]:
        """
        Get list of available tools.
        
        Returns:
            List of tool definitions
        """
        response = await self._request("GET", "/agent/tools")
        return response.json()
    
    async def add_extension(self, extension: Extension) -> str:
        """
        Add an extension to the agent.
        
        Args:
            extension: Extension configuration
            
        Returns:
            Success message
        """
        payload = {
            "name": extension.name,
            "type": extension.type
        }
        if extension.cmd:
            payload["cmd"] = extension.cmd
        if extension.args:
            payload["args"] = extension.args
        if extension.envs:
            payload["envs"] = extension.envs
        if extension.timeout:
            payload["timeout"] = extension.timeout
        if extension.bundled is not None:
            payload["bundled"] = extension.bundled
        
        response = await self._request("POST", "/agent/add_extension", json=payload)
        return response.text
    
    async def remove_extension(self, name: str) -> str:
        """
        Remove an extension from the agent.
        
        Args:
            name: Extension name
            
        Returns:
            Success message
        """
        response = await self._request(
            "POST",
            "/agent/remove_extension",
            json={"name": name}
        )
        return response.text
    
    async def list_extensions(self) -> List[Dict[str, Any]]:
        """
        List all available extensions.
        
        Returns:
            List of extension configurations
        """
        response = await self._request("GET", "/config/extensions")
        return response.json()
    
    # Session Management
    
    async def list_sessions(self) -> List[Session]:
        """
        List all sessions.
        
        Returns:
            List of Session objects
        """
        response = await self._request("GET", "/sessions")
        data = response.json()
        
        sessions = []
        for session_data in data.get("sessions", []):
            sessions.append(Session(
                id=session_data["id"],
                name=session_data.get("name"),
                created=session_data.get("created"),
                updated=session_data.get("updated"),
                working_dir=session_data.get("working_dir")
            ))
        return sessions
    
    async def get_session(self, session_id: str) -> Dict[str, Any]:
        """
        Get detailed session information.
        
        Args:
            session_id: Session ID
            
        Returns:
            Session details including conversation history
        """
        response = await self._request("GET", f"/sessions/{session_id}")
        return response.json()
    
    async def update_session_name(self, session_id: str, name: str) -> None:
        """
        Update a session's name.
        
        Args:
            session_id: Session ID
            name: New session name
        """
        await self._request(
            "POST",
            f"/sessions/{session_id}/name",
            json={"name": name}
        )
    
    async def delete_session(self, session_id: str) -> None:
        """
        Delete a session.
        
        Args:
            session_id: Session ID
        """
        await self._request("DELETE", f"/sessions/{session_id}")
    
    async def export_session(self, session_id: str) -> Dict[str, Any]:
        """
        Export a session's data.
        
        Args:
            session_id: Session ID
            
        Returns:
            Exported session data
        """
        response = await self._request("GET", f"/sessions/{session_id}/export")
        return response.json()
    
    async def get_session_insights(self) -> Dict[str, Any]:
        """
        Get insights about all sessions (token usage, costs, etc.).
        
        Returns:
            Session insights data
        """
        response = await self._request("GET", "/sessions/insights")
        return response.json()
    
    # Configuration
    
    async def get_providers(self) -> List[Dict[str, Any]]:
        """
        Get list of available LLM providers.
        
        Returns:
            List of provider configurations
        """
        response = await self._request("GET", "/config/providers")
        return response.json()
    
    async def get_provider_models(self, provider: str) -> List[str]:
        """
        Get available models for a provider.
        
        Args:
            provider: Provider name
            
        Returns:
            List of model names
        """
        response = await self._request("GET", f"/config/providers/{provider}/models")
        return response.json()
    
    async def check_status(self) -> Dict[str, Any]:
        """
        Check server status.
        
        Returns:
            Status information
        """
        response = await self._request("GET", "/status")
        return response.json()


# Example usage
async def example():
    """Example usage of the goose SDK."""
    async with GooseClient() as client:
        # Check status
        status = await client.check_status()
        print(f"Server status: {status}")
        
        # Start agent
        print("\nStarting agent...")
        await client.start_agent(provider="openai", model="gpt-4")
        
        # List available extensions
        extensions = await client.list_extensions()
        print(f"\nAvailable extensions: {len(extensions)}")
        
        # Send message
        print("\nSending message...")
        session_id = "example-session"
        
        async for event in client.send_message(
            session_id,
            "Write a simple hello world Python script"
        ):
            if event.type == "text" and event.message:
                for content in event.message.content:
                    if content.type == "text":
                        print(f"\nAgent: {content.text}")
            elif event.type == "Finish":
                print("\nConversation finished")


if __name__ == "__main__":
    asyncio.run(example())
